## Building a 331 Data Pipeline

    Book wants me to use nifi but I dont want to, gonna go on with apache airflow

https://seeclickfix.com/open311/v2/requests.json?lat=37.352390&long=-121.953079
https://seeclickfix.com/api/v2/issues?santa-clara-county

In [136]:
import pandas as pd
import requests
import json

In [137]:
param = {'place_url':'santa-clara-county-CA','per_page':'100'}
url = 'https://seeclickfix.com/open311/v2/requests.json?lat=37.3541&long=-121.9552'
r = requests.get(url).content

    the response is in form of python dictionary but it looks like json

In [138]:
b = r.decode('utf-8')

In [139]:
data = json.loads(b)

In [143]:
print(data)

[{'description': 'No Cody', 'address': '1675 Lincoln St Santa Clara CA 95050, الولايات المتحدة', 'lat': 37.3540231601478, 'service_request_id': 11742885, 'requested_datetime': '2022-01-15T10:01:35-05:00', 'long': -121.954946997951, 'service_code': None, 'status': 'open', 'zipcode': '95050', 'updated_datetime': '2022-01-15T10:51:47-05:00', 'agency_responsible': None, 'service_name': None, 'media_url': 'https://seeclickfix.com/rails/active_storage/representations/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBK1ZyV3c9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--a842c83091a4fbeda566eca6c303a0486716ddb6/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdDVG9MWm05eWJXRjBTU0lJYW5CbkJqb0dSVlE2RUdGMWRHOWZiM0pwWlc1MFZEb0tjM1J5YVhCVU9ndHlaWE5wZW1WSklndzRNREI0TmpBd0Jqc0dWQT09IiwiZXhwIjpudWxsLCJwdXIiOiJ2YXJpYXRpb24ifX0=--ee2a5b821fe0552ba6053903421e37a595d06845/image.jpg', 'expected_datetime': None, 'address_id': None, 'service_notice': None, 'status_notes': None}, {'description': ' There are about 20 potholes on 

In [144]:
df = pd.DataFrame(data).reset_index()
df.head()

index                                        description  \
0      0                                            No Cody   
1      1   There are about 20 potholes on Montague Expre...   
2      2  Huge pothole on 280 south bound before 85 cros...   
3      3  On 3rd street just west of the cross street Sa...   
4      4                   Test Description for santa clara   

                                             address        lat  \
0  1675 Lincoln St Santa Clara CA 95050, الولايات...  37.354023   
1  Montague Expressway between 880 & river oaks p...  37.354108   
2  Highway 280 Southbound And 85 Intersection San...  37.354108   
3  3rd Street Just West Of Santa Clara Alameda, C...  37.354108   
4     1690-1708 Triton Ct Santa Clara, CA 95050, USA  37.354296   

   service_request_id         requested_datetime        long service_code  \
0            11742885  2022-01-15T10:01:35-05:00 -121.954947         None   
1             2294004  2016-03-08T19:29:25-05:00 -121.955236         None   
2             1652038  2015-05-15T23:15:18-04:00 -121.955236         None   
3             1254599  2014-08-25T01:58:06-04:00 -121.955236         None   
4               72027  2011-01-06T04:57:48-05:00 -121.956785         None   

   status zipcode           updated_datetime agency_responsible service_name  \
0    open   95050  2022-01-15T10:51:47-05:00               None         None   
1  closed   95050  2016-07-07T18:53:06-04:00               None         None   
2  closed   95050  2015-11-12T11:29:05-05:00               None         None   
3    open   95050  2022-01-15T10:20:11-05:00               None         None   
4    open   95050  2022-01-15T10:20:17-05:00               None         None   

                                           media_url expected_datetime  \
0  https://seeclickfix.com/rails/active_storage/r...              None   
1                                               None              None   
2                                               None              None   
3                                               None              None   
4                                               None              None   

  address_id service_notice status_notes  
0       None           None         None  
1       None           None         None  
2       None           None         None  
3       None           None         None  
4       None           None         None

In [146]:
date = df['requested_datetime'].str.split(pat='T',n=1,expand=True)

In [147]:
df['date']=date[0]

In [151]:
use_df = df[['service_request_id','description','zipcode','lat','long','address','date']]
use_df.head()

service_request_id                                        description  \
0            11742885                                            No Cody   
1             2294004   There are about 20 potholes on Montague Expre...   
2             1652038  Huge pothole on 280 south bound before 85 cros...   
3             1254599  On 3rd street just west of the cross street Sa...   
4               72027                   Test Description for santa clara   

  zipcode        lat        long  \
0   95050  37.354023 -121.954947   
1   95050  37.354108 -121.955236   
2   95050  37.354108 -121.955236   
3   95050  37.354108 -121.955236   
4   95050  37.354296 -121.956785   

                                             address        date  
0  1675 Lincoln St Santa Clara CA 95050, الولايات...  2022-01-15  
1  Montague Expressway between 880 & river oaks p...  2016-03-08  
2  Highway 280 Southbound And 85 Intersection San...  2015-05-15  
3  3rd Street Just West Of Santa Clara Alameda, C...  2014-08-25  
4     1690-1708 Triton Ct Santa Clara, CA 95050, USA  2011-01-06

In [152]:
df.to_csv('issues.csv')

In [153]:
try:
    data = pd.read_csv('issues.csv')
except IOError:
    print("Trouble opening file")

In [154]:
data.head(2)

Unnamed: 0  index                                        description  \
0           0      0                                            No Cody   
1           1      1   There are about 20 potholes on Montague Expre...   

                                             address        lat  \
0  1675 Lincoln St Santa Clara CA 95050, الولايات...  37.354023   
1  Montague Expressway between 880 & river oaks p...  37.354108   

   service_request_id         requested_datetime        long  service_code  \
0            11742885  2022-01-15T10:01:35-05:00 -121.954947           NaN   
1             2294004  2016-03-08T19:29:25-05:00 -121.955236           NaN   

   status  zipcode           updated_datetime  agency_responsible  \
0    open    95050  2022-01-15T10:51:47-05:00                 NaN   
1  closed    95050  2016-07-07T18:53:06-04:00                 NaN   

   service_name                                          media_url  \
0           NaN  https://seeclickfix.com/rails/active_storage/r...   
1           NaN                                                NaN   

   expected_datetime  address_id  service_notice  status_notes        date  
0                NaN         NaN             NaN           NaN  2022-01-15  
1                NaN         NaN             NaN           NaN  2016-03-08

In [76]:
date = data['created_at'].str.split(pat='T',n=1,expand=True)
data['date']=date[0]
use_df = data[['id','summary','description','rating','lat','lng','address','date']]
use_df.head()

id                                  summary  \
0  13699499                  Public Restrooms Issues   
1  13699498  Litter & Illegal Dumping (Public Works)   
2  13699497                        Homeless Concerns   
3  13699496                                 Homeless   
4  13699495                             Street Light   

                                         description  rating        lat  \
0  The faucet in the men’s locker room in the poo...       1  38.857980   
1  dumped mattress and other items.  shopping car...       2  48.762009   
2  Large homeless camp on the hill on the south s...       2  32.686449   
3      Homeless tent in front of Kehalani McDonald's       1  20.872751   
4  A streetlight is out between 20411 and 20451 W...       1  49.170044   

          lng                                            address        date  
0 -104.898004  202 Manitou Ave Manitou Springs CO 80829, Unit...  2022-11-16  
1 -122.501937        2400 Keesling St Bellingham, WA, 98225, USA  2022-11-16  
2 -117.070557  3600-3686 Paradise Valley Rd National City, CA...  2022-11-16  
3 -156.500159     105 Kehalani Village Dr Wailuku, HI 96793, USA  2022-11-16  
4 -122.995171  20411 Westminster Hwy N Richmond, BC, V6V 1B3,...  2022-11-16

insert into seeClickFix(id, summary, description, rating, lat, long, address, date) values(13699499,'Public Restrooms Issues','dasdf',1,38.857980,-104.898004,'dfasdf','2022-11-16');

In [155]:
use_df.to_csv('filtered.csv')

In [156]:
import psycopg2 as db
['service_request_id','description','zipcode','lat','long','address','date']

In [133]:
try:
        con = "dbname = 'postgres' host='192.168.1.41' user='postgres' password=''"
        connection = db.connect(con)
        cur=connection.cursor()
except:
    print("Trouble connecting to postgres server")
try:
    df = pd.read_csv('filtered.csv')
except IOError:
    print("Trouble opening file")
cur.execute(query='create table if not exists seeClickFix(id bigint primary key, description text, zipcode int, lat double precision, long double precision, address text, date date)')
    

In [134]:
for i,r in df.iterrows():
        id = r['id']
        summary = r['summary']
        desc = r['description']
        desc = desc.replace('\'','')
        rat = r['rating']
        lat = r['lat']
        lng = r['lng']
        add = r['address']
        dat = r['date']
        cur.execute(query=f"insert into seeClickFix(id, summary, description, rating, lat, long, address, date) values({id},'{summary}','{desc}',{rat},{lat},{lng},'{add}','{dat}');")

In [135]:
connection.commit()
cur.close()
connection.close()

In [104]:
text = 'Homeless tent in front of Kehalani McDonald\'s'
print(text)

Homeless tent in front of Kehalani McDonald's


In [ ]:
cur.execute(query=f"insert into seeClickFix(id, summary, description, rating, lat, long, address, date) values({id},'{summary}'','{desc}',{rat},{lat},{lng},'{add}','{dat}');")